### PR TITLE
Move a bunch of stuff out of `internal_macros`

### PR DESCRIPTION
**Pull request by tcharding**
Move macros out of `internal_macros`, done in an effort to work towards removing the `internal_macros` module since we have `bitcoin_internals` now.

### CODE REVIEW

**Comment by apoelstra** [comment]
concept ACK everything except the `consensus_encoding` stuff. As you say, we don't know the final home for these, but my inclination is that it's best to leave the malone for now.

**Comment by tcharding** [comment]
> concept ACK everything except the consensus_encoding stuff

Removed `consensus_encoding` changes that were present in the draft of this PR.

**Review by apoelstra — DISMISSED** [review]
ACK 5b0afa7d7a0b6a70fb6ed48ad1d4d87ecbe36ccb

**Comment by tcharding** [comment]
Rebase only, no other changes.

**Review by apoelstra — APPROVED** [review]
ACK 3f2fc51ad9a4499ca3cf8917fa1e033bb4d8edb7

**Review by apoelstra — DISMISSED** [review]
ACK 8bad993d556fcfcd32b6e884bab302461a56cc16

**Review by sanket1729 — DISMISSED** [review]
ACK 8bad993d556fcfcd32b6e884bab302461a56cc16. My comment is not relevant to this PR, but we should address it later

File: internals/src/macros.rs [hunk 1]
```diff
@@ -1,42 +1,42 @@
...
             #[inline]
             pub fn is_empty(&self) -> bool { false }
 
-            /// Returns the underlying bytes.
+            /// Returns a reference the underlying bytes.
             #[inline]
             pub fn as_bytes(&self) -> &[$ty; $len] { &self.0 }
 
-            /// Returns the underlying bytes.
+            /// Returns a clone of the underlying bytes.
```
[the remark this answers is not in the document]
Comment (tcharding) [reply internals/src/macros.rs:35]: Good point, `into_bytes` can be removed. Added to my todo list.

**Comment by tcharding** [comment]
Rebase only, no other changes.

**Review by apoelstra — APPROVED** [review]
ACK 7d851b42ee2b5b2183e9ff6d5dc3b57a78dd943f

File: internals/src/macros.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+//!
+
+/// Implements standard array methods for a given wrapper type
+#[macro_export]
+macro_rules! impl_array_newtype {
+    ($thing:ident, $ty:ty, $len:literal) => {
+        impl $thing {
+            /// Converts the object to a raw pointer
+            #[inline]
+            pub fn as_ptr(&self) -> *const $ty {
```
Comment (Kixunil) [inline internals/src/macros.rs:11]: I just realized this is effectively `unsafe` code. We don't explicitly say that `unsafe` code can rely on the validity of the returned pointer so it's till `unsafe` consumer to blame if we get it wrong but without guaranteeing validity the method is near useless - you can at best use it to print out the pointer. And `as_mut_ptr` is thus redundant.

File: internals/src/macros.rs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                let &$thing(ref dat) = self;
```
Comment (Kixunil) [inline internals/src/macros.rs:12]: If we decide to guarantee validity `let inner: &[$ty; $len] = &self.0; inner as *const $ty` would be safer.

File: internals/src/macros.rs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                dat.as_mut_ptr()
+            }
+
+            /// Returns the length of the object as an array
+            #[inline]
+            pub fn len(&self) -> usize { $len }
+
+            /// Returns whether the object, as an array, is empty. Always false.
+            #[inline]
+            pub fn is_empty(&self) -> bool { false }
```
Comment (Kixunil) [inline internals/src/macros.rs:29]: If this is just to make `clippy` shut up I'd rather `#[allow]` the lint.

File: internals/src/macros.rs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+            pub fn to_bytes(self) -> [$ty; $len] { self.0.clone() }
+
+            /// Returns the underlying bytes.
+            #[inline]
+            pub fn into_bytes(self) -> [$ty; $len] { self.0 }
+        }
+
+        impl<'a> core::convert::From<&'a [$ty]> for $thing {
+            fn from(data: &'a [$ty]) -> $thing {
+                assert_eq!(data.len(), $len);
```
Comment (Kixunil) [inline internals/src/macros.rs:46]: NO, NO, NO, NO, NO!!!! This is completely wrong implementation of `From`. Should've been `TryFrom` or `From<[$ty; $len]>`

File: internals/src/macros.rs [hunk 6]
```diff
@@ -1,42 +1,42 @@
...
+            /// Returns the length of the object as an array.
             #[inline]
             pub fn len(&self) -> usize { $len }
 
             /// Returns whether the object, as an array, is empty. Always false.
             #[inline]
             pub fn is_empty(&self) -> bool { false }
 
-            /// Returns the underlying bytes.
+            /// Returns a reference the underlying bytes.
```
Comment (Kixunil) [inline internals/src/macros.rs:31]: reference *to* the underlying bytes

File: internals/src/macros.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kixunil) [reply internals/src/macros.rs:35]: In that case use `self.0` in the impl to guard against calling on non-`Copy` types.

File: internals/src/macros.rs [hunk 7]
```diff
@@ -73,3 +73,11 @@ macro_rules! debug_from_display {
         }
     };
 }
+
+/// Asserts a boolean expression at compile time.
+#[macro_export]
+macro_rules! const_assert {
+    ($x:expr) => {{
+        const _: [(); 0 - !$x as usize] = [];
```
Comment (Kixunil) [inline internals/src/macros.rs:81]: Nit: I didn't realize it before but I find `const _: () = [][!$x as usize];` more readable

**Review by Kixunil — APPROVED** [review]
ACK 7d851b42ee2b5b2183e9ff6d5dc3b57a78dd943f

The doc can be improved in followup PR. Also invalid `From` impl is unrelated to moving.

**Comment by tcharding** [comment]
Thanks for the review @Kixunil, I've added "do follow up to this PR" to my todo list.

File: internals/src/macros.rs (hunk 3 above)
Comment (apoelstra) [reply internals/src/macros.rs:12]: I agree your code is more obviously correct, but is the existing code wrong?

**Comment by apoelstra** [comment]
I'm going to go ahead and merge this, since this move-only PR isn't the place to fix issues with this macro.

File: internals/src/macros.rs (hunk 3 above)
Comment (Kixunil) [reply internals/src/macros.rs:12]: The macro doesn't define the type so supplying a wrong value to the macro could make this unsound. Although it's caught by `to_bytes()` method which wouldn't type check. I'd rather not rely on things far a way from this code - the main benefit of `unsafe` is being able to limit the scope of affected code.

**Pull request merged**